src/safeposix/syscalls/net_calls.rs

```diff
@@ -1234,4 +1233,274 @@ impl Cage {
+                        }
+                        //if the fd already exists, insert overwrites the prev entry
+                        epollfdobj.registered_fds.insert(fd, EpollEvent { events: event.events, fd: event.fd });
+                    }
+                    EPOLL_CTL_ADD => {
+                        if let Some(_) = epollfdobj.registered_fds.get(&fd) {
+                            return syscall_error(Errno::EEXIST, "epoll ctl", "fd is already registered");
+                        }
+                        epollfdobj.registered_fds.insert(fd, EpollEvent { events: event.events, fd: event.fd });
+                    }
+                    _ => {
+                        return syscall_error(Errno::EINVAL, "epoll ctl", "provided op is invalid");
+                    }
+                }
+            } else {
+                return syscall_error(Errno::EBADF, "epoll ctl", "provided fd is not a valid file descriptor")
+            }
+        } else {
+            return syscall_error(Errno::EBADF, "epoll ctl", "provided epoll fd is not a valid epoll file descriptor");
+        }
+        return 0;
+    }
+
+    pub fn epoll_wait_syscall(&self, epfd: i32, maxevents: usize, timeout: Option<interface::RustDuration>) -> i32 {
+
+        let fdtable = self.filedescriptortable.read().unwrap();
+
+        if let Some(wrappedfd) = fdtable.get(&epfd) {
+            let filedesc_enum = wrappedfd.read().unwrap();
+            if let Epoll(epollfdobj) = &*filedesc_enum {
+                if !maxevents > 0 {
+                    return syscall_error(Errno::EINVAL, "epoll wait", "max events argument is not a positive number");
+                }
+
+                //now we know that all of the arguments are valid...
+                let mut reads = interface::RustHashSet::<i32>::new();
+                let mut writes = interface::RustHashSet::<i32>::new();
+                let mut errors = interface::RustHashSet::<i32>::new();
+
+                let mut poll_fds_vec: Vec<PollStruct> = vec![];
+
+                for (&key, &value) in &epollfdobj.registered_fds {
+
+                    let events = value.events;
+                    let mut structpoll = PollStruct {
+                        fd: key,
+                        events: 0,
+                        revents: 0
+                    };
+
+                    if events & EPOLLIN as u32 > 0 {
+                        structpoll.events |= POLLIN;
+                    }
+                    if events & EPOLLOUT as u32 > 0 {
+                        structpoll.events |= POLLOUT;
+                    }
+                    if events & EPOLLERR as u32 > 0 {
+                        structpoll.events |= POLLERR;
+                    }
+                    poll_fds_vec.push(structpoll);
+
+                }
+
+                let mut poll_fds_slice = &mut poll_fds_vec[..];
+                Self::poll_syscall(&self, &mut *poll_fds_slice, timeout);
```

Comment:
As far as I can tell, you shouldn't have to dereference and rereference here because you already have a mutable slice above?

Reply:
I might be wrong though.